spec-linter/spec_linter.go

```diff
@@ -0,0 +1,190 @@
+			fileList = append(fileList, path)
+		}
+
+		return nil
+	})
+	if err != nil {
+		panic(err)
+	}
+
+	for _, file := range fileList {
+		fmt.Printf("\n\n%s\n", file)
+		b, err := os.ReadFile(file)
+		if err != nil {
+			panic(err)
+		}
+		var anyJson map[string]interface{}
+		err = json.Unmarshal(b, &anyJson)
+		if err != nil {
+			panic(err)
+		}
+
+		// struct fields in the digital datasheet schema
+		keyList := maps.Keys(anyJson)
+
+		// each file should contain "$schema"
+		v, ok := anyJson["$schema"]
+		if !ok {
+			panic(fmt.Errorf("expecting $schema key in %v", keyList))
+		}
+
+		// $title should start with "specification"
+		v, ok = anyJson["title"]
+		if !ok {
+			panic(fmt.Errorf("expecting title key in %v", keyList))
+		}
+		title := fmt.Sprintf("%v", v)
+		if strings.HasPrefix(title, "Specification") {
+			anyJson["title"] = strings.ToLower(title)
+			fmt.Printf("fixed title: %s\n", anyJson["title"])
+		} else if !strings.HasPrefix(title, "specification") {
+			anyJson["title"] = fmt.Sprintf("specification of %s", title)
+			fmt.Printf("fixed title: %s\n", anyJson["title"])
+		}
+
+		// $id should have this format: "https://github.com/edatasheets/digital-datasheets/blob/main/part-spec/clock/oscillator.json"
+		v, ok = anyJson["$id"]
+		if !ok {
+			panic(fmt.Errorf("expecting $id key in %v", keyList))
+		}
+		idString := fmt.Sprintf("%v", v)
+		prefixPath := "https://github.com/edatasheets/digital-datasheets/blob/main/part-spec/"
+		correctPrefix := strings.HasPrefix(idString, prefixPath)
+		filePieces := strings.Split(file, "/")
+		suffixPath := filepath.Join(filePieces[len(filePieces)-2], filePieces[len(filePieces)-1])
+		correctPath := strings.HasSuffix(idString, suffixPath)
+		if !correctPrefix || !correctPath {
+			anyJson["$id"] = filepath.Join(prefixPath, suffixPath)
+			fmt.Printf("fixed $id: %s\n", anyJson["$id"])
+		}
+
+		// For each property key:
+		//		1. Property name should be lowercase
+		//		2. Description should be lowercase
+		//		3. Description should not end in a period
+		//		// TODO: this has currently been fixed by hand, but we should add this to the script
+		//		4. Ref should go to an actual place
+
+		// Drill into the structure to get to the component properties
+		objectName := strings.TrimSuffix(filepath.Base(file), ".json")
+		v, ok = anyJson[objectName]
+		if !ok {
+			fmt.Printf("expecting %s key in %v\n", objectName, keyList)
+			continue
+		}
+		vMap, err := castToMap(v)
+		if err != nil {
+			fmt.Println(err)
+			continue
+		}
+
+		v, ok = vMap["properties"]
+		if !ok {
+			fmt.Printf("expecting properties key in %v\n", maps.Keys(vMap))
+		}
+		vMap, err = castToMap(v)
+		if err != nil {
+			fmt.Println(err)
+			continue
+		}
+
+		for k, v := range vMap {
+			// make sure the property names start with a lowercase letter
+			if checkForCapitalWithSkips(k) {
+				newK := strings.ToLower(string(k[0])) + k[1:]
+				vMap[newK] = v
+				delete(vMap, k)
```

Comment:
I think it would be simpler to delete from vMap before setting the new one: if there is some weird bug (now or in the future) where `k == newK`, this code would cause the vMap to go empty whereas if delete happened before updating `vMap[newK]`, having `k == newK` would be immaterial.

Since this is a validation script, we really want to avoid bugs which cause the vMap to lose entries.

Reply:
good point